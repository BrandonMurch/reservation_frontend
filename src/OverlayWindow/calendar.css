/* Must use !important to ensure that the FullCalendar css is overridden */

td.fc-day-past {
    background-color: transparent !important;
    color: rgba(150,150,150,0.5);
    cursor: default;
}

td.fc-day-today {
    background-color: rgba(255,255,255,0.5) !important;
    color: black !important;
}

td.fc-day-disabled {
    background-color: transparent !important;
    cursor: default;
}
