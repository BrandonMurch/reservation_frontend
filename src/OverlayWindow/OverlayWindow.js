import React from 'react';
import FullCalendar from '@fullcalendar/react';
import dayGridPlugin from '@fullcalendar/daygrid'
import style from './OverlayWindow.module.css';
import './calendar.css';

class OverlayWindow extends React.Component {

    getJSX() {
        return (
            <div className={style.overlay}>
                <ExitOverlay closeOverlay={this.props.closeOverlay}/>
                <Calendar />
            </div>
        )
    }

    render() {
        const display = this.props.hidden ? "" : this.getJSX();
        return (
            <div>
                {display}
            </div>
        )
    }
}

class Calendar extends React.Component{

    getValidRange() {

        var startDate = new Date();
        var endDate = new Date();
        endDate.setFullYear(endDate.getFullYear() + 1);

        startDate = startDate.toISOString().slice(0, 10);
        endDate = endDate.toISOString().slice(0, 10);

        return {start: startDate, end: endDate};
    }

    render() {
        return (
            <div className={style.calendar}>
                <FullCalendar
                    plugins={[ dayGridPlugin ]}
                    initialView="dayGridMonth"
                    showNonCurrentDates={false}
                    fixedWeekCount={false}
                    height="100%"
                    validRange={this.getValidRange()}
                    // viewDidMount={(currentView) => {
                    //     const today = new Date();
                    //     console.log(currentView);
                    //     console.log(currentView.currentStart);
                    //     console.log(FullCalendar.view)
                    //     if (today > currentView.currentStart && today <= currentView.currentEnd) {
                    //         var prevButton = document.getElementsByClass(".fc-prev-button");
                    //         prevButton.forEach((entry) => {
                    //             entry.disabled = true;
                    //             entry.addClass('fc-state-disabled')
                    //         })
                    //     }
                    // }}
                />
            </div>
        )
    }
}

function ExitOverlay(props) {
    return (
        <div className={style.exitContainer}>
            <p
            className={style.exitText}
            onClick={props.closeOverlay}>
                x
            </p>
        </div>
    )
}

export default OverlayWindow;
