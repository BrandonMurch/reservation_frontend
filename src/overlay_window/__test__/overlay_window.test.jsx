// Dependencies
import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { fireEvent } from '@testing-library/react';
import { unmountComponentAtNode } from 'react-dom';
import { create } from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { renderWithRouter, mockFetch } from 'test_utils';

// Components
import OverlayWindow from '../index';

const mockCalendarFetch = function mockFetchResponseFromServer() {
  const today = new Date();
  const todayDate = today.getDate();
  const todayString = today.toLocaleDateString('en-CA');
  const tomorrowString = todayString.replace(new RegExp(`${todayDate}$`, 'i'), todayDate + 1);
  const endDayString = todayString.replace(new RegExp(`${todayDate}$`, 'i'), todayDate + 3);

  const mockSuccessResponse = {
    start: todayString,
    end: endDayString,
    availableDates: [todayString, tomorrowString],
  };
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  return Promise.resolve({
    json: () => mockJsonPromise,
  });
};

describe('<OverlayWindow />', () => {
  let component;
  let mockCloseOverlayFunction;
  let fetchSpy;
  let container = null;
  const reservation = {
    date: '2020-10-10',
    time: '21:00',
    partySize: 2,
  };
  const user = {
    firstName: 'john',
    lastName: 'johnson',
    email: 'john@john.com',
    phoneNumber: '+1 123456787',
    tAC: true,
  };
  beforeEach(async () => {
    container = document.createElement('div');
    document.body.appendChild(container);
    mockCloseOverlayFunction = jest.fn();
  });

  afterEach(() => {
    if (fetchSpy) {
      fetchSpy.mockClear();
    }
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('should match snapshot', () => {
    const tree = create(
      <Router>
        <OverlayWindow closeOverlay={mockCloseOverlayFunction} />
      </Router>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should submit reservation to server', async () => {
    fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => mockFetch(201));

    component = await renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/review' },
    );
    const submit = component.getByRole('button', { name: 'Make reservation.' });
    await act(async () => {
      await fireEvent.click(submit);
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(component.history.location.pathname).toEqual('/success');
  });

  it('should setError on server error', async () => {
    fetchSpy = jest
      .spyOn(global, 'fetch')
      .mockImplementation(() => mockFetch(409, { message: 'Error!' }));

    component = renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/review' },
    );
    const submit = component.getByRole('button', { name: 'Make reservation.' });

    await act(async () => {
      await fireEvent.click(submit);
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const errorText = component.getByText('Error!');
    expect(errorText).toBeInTheDocument();
  });

  it('should setError on fetch error', async () => {
    fetchSpy = jest
      .spyOn(global, 'fetch')
      .mockImplementation(() => Promise.reject(new Error('Error!')));

    component = await renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/review' },
    );
    const submit = component.getByRole('button', { name: 'Make reservation.' });
    await act(async () => {
      await fireEvent.click(submit);
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    const errorText = component.getByText('Error!');
    expect(errorText).toBeInTheDocument();
  });

  it('display a link to close the overlay', () => {
    component = renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/review' },
    );
    const exitLink = component.getByRole('link');
    expect(exitLink).toBeInTheDocument();
    act(() => {
      fireEvent.click(exitLink);
    });

    expect(component.history.location.pathname).toEqual('/');
    expect(mockCloseOverlayFunction).toHaveBeenCalledTimes(1);
  });

  it('should display calendar on /calendar', async () => {
    fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => mockCalendarFetch());
    await act(async () => {
      component = await renderWithRouter(
        <OverlayWindow
          closeOverlay={mockCloseOverlayFunction}
          reservationInfo={reservation}
          userInfo={user}
        />,
        { route: '/calendar' },
      );
    });
    let text = component.getByText('Wed');
    expect(text).toBeInTheDocument();
    text = component.getByText('Fri');
    expect(text).toBeInTheDocument();
  });

  it('should display reservation on /reservation', () => {
    component = renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/reservation' },
    );

    const comboBoxes = component.getAllByRole('combobox');
    expect(comboBoxes).toHaveLength(2);
    const text = component.getByText(/Desired date/i);
    expect(text).toBeInTheDocument();
  });

  it('should display review on /review', () => {
    component = renderWithRouter(
      <OverlayWindow
        closeOverlay={mockCloseOverlayFunction}
        reservationInfo={reservation}
        userInfo={user}
      />,
      { route: '/review' },
    );
    const buttons = component.getAllByRole('button');
    expect(buttons).toHaveLength(4);
    const text = component.getByText(new RegExp(user.firstName));
    expect(text).toBeInTheDocument();
  });

  it('should display success on /success', () => {
    act(() => {
      component = renderWithRouter(
        <OverlayWindow
          closeOverlay={mockCloseOverlayFunction}
          reservationInfo={reservation}
          userInfo={user}
        />,
        { route: '/success' },
      );
    });

    const success = component.getByText(/Reservation was made successfully/i);
    expect(success).toBeInTheDocument();
  });
});
