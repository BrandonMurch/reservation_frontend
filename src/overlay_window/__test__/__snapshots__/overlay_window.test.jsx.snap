// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverlayWindow /> should match snapshot 1`] = `
<div
  className="overlay"
  data-testid="overlay-window"
>
  
  <div
    className="exitContainer"
  >
    <a
      className="exitText"
      href="/"
      onClick={[Function]}
    >
      x
    </a>
  </div>
</div>
`;
